worker/restore_map.go

```diff
@@ -0,0 +1,708 @@
+// values from predicates no longer assigned to this group.
+// If restoreTs is greater than zero, the key-value pairs will be written with that timestamp.
+// Otherwise, the original value is used.
+// TODO(DGRAPH-1234): Check whether restoreTs can be removed.
+func (m *mapper) Map(r io.Reader, in *loadBackupInput) error {
+	br := bufio.NewReaderSize(r, 16<<10)
+	zbuf := z.NewBuffer(bufSz, "Restore.Map")
+
+	for {
+		var sz uint64
+		err := binary.Read(br, binary.LittleEndian, &sz)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		m.szHist.Update(int64(sz))
+		buf := zbuf.SliceAllocate(int(sz))
+		if _, err = io.ReadFull(br, buf); err != nil {
+			return err
+		}
+
+		if zbuf.LenNoPadding() > bufSoftLimit {
+			atomic.AddUint64(&m.bytesRead, uint64(zbuf.LenNoPadding()))
+			glog.Infof("Sending req of size: %s\n", humanize.IBytes(uint64(zbuf.LenNoPadding())))
+			m.reqCh <- listReq{zbuf, in}
+			zbuf = z.NewBuffer(bufSz, "Restore.Map")
+		}
+	}
+	m.reqCh <- listReq{zbuf, in}
+	return nil
+}
+
+type mapResult struct {
+	maxUid uint64
+	maxNs  uint64
+}
+
+// 1. RunMapper creates a mapper object
+// 2. mapper.Map() ->
+func RunMapper(req *pb.RestoreRequest, mapDir string) (*mapResult, error) {
+	uri, err := url.Parse(req.Location)
+	if err != nil {
+		return nil, err
+	}
+	if req.RestoreTs == 0 {
+		return nil, errors.New("RestoreRequest must have a valid restoreTs")
+	}
+
+	creds := getCredentialsFromRestoreRequest(req)
+	h, err := NewUriHandler(uri, creds)
+	if err != nil {
+		return nil, err
+	}
+
+	manifests, err := getManifestsToRestore(h, uri, req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot retrieve manifests")
+	}
+	glog.Infof("Got %d backups to restore ", len(manifests))
+
+	cfg, err := getEncConfig(req)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get encryption config")
+	}
+	keys, err := ee.GetKeys(cfg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get encryption keys")
+	}
+
+	mapper := &mapper{
+		buf:       z.NewBuffer(mapFileSz, "Restore.Buffer"),
+		thr:       y.NewThrottle(3),
+		bufLock:   &sync.Mutex{},
+		closer:    z.NewCloser(1),
+		reqCh:     make(chan listReq, 3),
+		restoreTs: req.RestoreTs,
+		mapDir:    mapDir,
+		szHist:    z.NewHistogramData(z.HistogramBounds(10, 32)),
+	}
+
+	numGo := 8
+	g, ctx := errgroup.WithContext(mapper.closer.Ctx())
+	for i := 0; i < numGo; i++ {
+		g.Go(func() error {
+			return mapper.processReqCh(ctx)
+		})
+	}
+	go mapper.Progress()
+	defer func() {
+		mapper.Flush()
+		mapper.closer.SignalAndWait()
+	}()
+
+	dropAll := false
+	dropAttr := make(map[string]struct{})
+
+	// manifests are ordered as: latest..full
+	for i, manifest := range manifests {
+		// A dropAll or DropData operation is encountered. No need to restore previous backups.
+		if dropAll {
+			break
+		}
+		if manifest.ValidReadTs() == 0 || len(manifest.Groups) == 0 {
```

Comment:
What can be the cases we dont' have a valid read ts? Is it just backwards compatibility thing? Can you add a comment explaining.

Reply:
just covering edge cases

Reply:
just covering corner cases I believe.